File: flask_app/app/dao/notification_dao.py
from app.models.notifications import Notifications
from app.models import db
from sqlalchemy import and_
from datetime import datetime

class NotificationDAO:

    @staticmethod
    def get_all_for_user(user_id, page, per_page, filters=None):
        query = Notifications.query.filter_by(user_id=user_id)

        if filters:
            if filters.get("is_read") is not None:
                query = query.filter(Notifications.is_read == filters["is_read"])

            if filters.get("notification_type"):
                query = query.filter(Notifications.notification_type == filters["notification_type"])

            if filters.get("start_date"):
                query = query.filter(Notifications.created_at >= filters["start_date"])

            if filters.get("end_date"):
                query = query.filter(Notifications.created_at <= filters["end_date"])

        pagination = query.paginate(page=page, per_page=per_page, error_out=False)
        return pagination.items, pagination.total

    @staticmethod
    def get_by_id(notification_id):
        return Notifications.query.filter_by(id=notification_id).first()

    @staticmethod
    def create(data):
        notification = Notifications(**data)
        db.session.add(notification)
        db.session.commit()
        return notification

    @staticmethod
    def mark_as_read(notification_id):
        notification = NotificationDAO.get_by_id(notification_id)
        if not notification:
            return None
        notification.is_read = True
        db.session.commit()
        return notification

    @staticmethod
    def mark_all_as_read(user_id):
        Notifications.query.filter_by(user_id=user_id, is_read=False).update({Notifications.is_read: True})
        db.session.commit()

File: springboot_api/src/main/java/com/example/membership_api/dao/NotificationDAO.java
package com.example.membership_api.dao;

import com.example.membership_api.model.Notifications;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import jakarta.persistence.TypedQuery;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Map;

@Repository
public class NotificationDAO {

    @PersistenceContext
    private EntityManager em;

    public List<Notifications> getAllForUser(String userId, int page, int perPage, Map<String, Object> filters) {
        StringBuilder queryBuilder = new StringBuilder("SELECT n FROM Notifications n WHERE n.userId = :userId");

        if (filters != null) {
            if (filters.containsKey("is_read")) {
                queryBuilder.append(" AND n.isRead = :isRead");
            }
            if (filters.containsKey("notification_type")) {
                queryBuilder.append(" AND n.notificationType = :notificationType");
            }
            if (filters.containsKey("start_date")) {
                queryBuilder.append(" AND n.createdAt >= :startDate");
            }
            if (filters.containsKey("end_date")) {
                queryBuilder.append(" AND n.createdAt <= :endDate");
            }
        }

        queryBuilder.append(" ORDER BY n.createdAt DESC");

        TypedQuery<Notifications> query = em.createQuery(queryBuilder.toString(), Notifications.class);
        query.setParameter("userId", userId);

        if (filters != null) {
            if (filters.containsKey("is_read")) {
                query.setParameter("isRead", (Boolean) filters.get("is_read"));
            }
            if (filters.containsKey("notification_type")) {
                query.setParameter("notificationType", filters.get("notification_type"));
            }
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
            if (filters.containsKey("start_date")) {
                query.setParameter("startDate", LocalDate.parse((String) filters.get("start_date"), formatter).atStartOfDay());
            }
            if (filters.containsKey("end_date")) {
                query.setParameter("endDate", LocalDate.parse((String) filters.get("end_date"), formatter).atTime(23, 59, 59));
            }
        }

        query.setFirstResult((page - 1) * perPage);
        query.setMaxResults(perPage);

        return query.getResultList();
    }

    public long countForUser(String userId, Map<String, Object> filters) {
        StringBuilder queryBuilder = new StringBuilder("SELECT COUNT(n) FROM Notifications n WHERE n.userId = :userId");

        if (filters != null) {
            if (filters.containsKey("is_read")) {
                queryBuilder.append(" AND n.isRead = :isRead");
            }
            if (filters.containsKey("notification_type")) {
                queryBuilder.append(" AND n.notificationType = :notificationType");
            }
            if (filters.containsKey("start_date")) {
                queryBuilder.append(" AND n.createdAt >= :startDate");
            }
            if (filters.containsKey("end_date")) {
                queryBuilder.append(" AND n.createdAt <= :endDate");
            }
        }

        TypedQuery<Long> query = em.createQuery(queryBuilder.toString(), Long.class);
        query.setParameter("userId", userId);

        if (filters != null) {
            if (filters.containsKey("is_read")) {
                query.setParameter("isRead", (Boolean) filters.get("is_read"));
            }
            if (filters.containsKey("notification_type")) {
                query.setParameter("notificationType", filters.get("notification_type"));
            }
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
            if (filters.containsKey("start_date")) {
                query.setParameter("startDate", LocalDate.parse((String) filters.get("start_date"), formatter).atStartOfDay());
            }
            if (filters.containsKey("end_date")) {
                query.setParameter("endDate", LocalDate.parse((String) filters.get("end_date"), formatter).atTime(23, 59, 59));
            }
        }

        return query.getSingleResult();
    }

    public Notifications getById(String notificationId) {
        return em.find(Notifications.class, notificationId);
    }

    @Transactional
    public Notifications create(Notifications notification) {
        em.persist(notification);
        return notification;
    }

    @Transactional
    public Notifications markAsRead(String notificationId) {
        Notifications notification = getById(notificationId);
        if (notification == null) {
            return null;
        }
        notification.setRead(true);
        em.merge(notification);
        return notification;
    }

    @Transactional
    public void markAllAsRead(String userId) {
        em.createQuery("UPDATE Notifications n SET n.isRead = true WHERE n.userId = :userId AND n.isRead = false")
          .setParameter("userId", userId)
          .executeUpdate();
    }
}
